README.md
# MemHC

MemHC is a high performance multithreaded event-based key/value cache
store, modified based on Memcached. MemHC use Hill-Cache cache replacement policy to help the system achieve better hit rate results.

## Dependencies

* libevent - https://www.monkey.org/~provos/libevent/ (libevent-dev)
* libseccomp (optional, experimental, linux) - enables process restrictions for
  better security. Tested only on x86-64 architectures.
* openssl (optional) - enables TLS support. need relatively up to date
  version. pkg-config is needed to find openssl dependencies (such as -lz).


## Build
```
./configure --prefix=/path/to/memhc
make -j$(nproc) && make install
```

## Environment

Be warned that the -k (mlockall) option to memcached might be
dangerous when using a large cache. Just make sure the memcached machines
don't swap.  memcached does non-blocking network I/O, but not disk.  (it
should never go to disk, or you've lost the whole point of it)

## Build status

See https://build.memcached.org/ for multi-platform regression testing status.

## Bug reports

Feel free to use the issue tracker on github.

**If you are reporting a security bug** please contact a maintainer privately.
We follow responsible disclosure: we handle reports privately, prepare a
patch, allow notifications to vendor lists. Then we push a fix release and your
bug can be posted publicly with credit in our release notes and commit
history.

## Website

* https://www.memcached.org

## Contributing

See https://github.com/memcached/memcached/wiki/DevelopmentRepos


ChangeLog
2016-08-23
    * ChangeLog moved from Google Code to Github Wiki, it's now available at:
      https://github.com/memcached/memcached/wiki/ReleaseNotes

2010-10-11
	* ChangeLog is no longer being updated.
	See http://code.google.com/p/memcached/wiki/ReleaseNotes

2009-04-10

	* *critical bugfix*. In 1.2.7 under multithreaded mode, memcached
	  would never restart accepting connections after hitting the
	  maximum connection limit.

2009-04-02

	* Fix: Rewrote stat buffer handling (Trond)

2009-03-31

	* Feature:  -b option for configuring backlog

	* Feature:  stats settings

	* Fix: binary stat alignment bug (bug26)

	* Fix: buffer overflow in stats (bug27)

	* Fix: recycle memory more aggressively (bug14)

	* Fix: incr validation (bug31)

	* Fix: 64-bit incr/decr delta problem (bug21)

	* Fix: Ascii UDP set (bug36)

	* Fix: stats slabs' used chunks (bug29)

	* Fix: stats reset resetting more (bug22)

	* Misc: More tests, documentation, cleanliness (godliness)

	* Stable merge (stats, debuggability, listen bugs)

2009-03-11

	* Protocol:  Binary complete (Dustin, Trond, Toru, etc...)

	* Performance:  improvements from facebook (Dormando)

	* Performance:  per-thread stats (Trond)

	* Performance:  Hash expansion in its own thread (Trond)

	* Stats:  Lots of new stats (Dustin)

	* Bug fixes (various)

	See the 1.3.2 release notes or revision control for more detail.

2008-09-06
       * Display error status on listen failures (Dormando)

       * Remove managed instance code. Incomplete/etc. (Dormando)

          * Handle broken IPV6 stacks better (Brian Aker)

          * Generate warnings on setsockopt() failures (Brian Aker)

          * Fix some indentation issues (Brian Aker)

          * UDP/TCP can be disabled by setting their port to zero (Brian Aker)

          * Zero out libevent thread structures before use (Ricky Zhou)

          * New stat: Last accessed time for last evicted item per slab class.
            (Dormando)

          * Use a dedicated socket accept thread (Facebook)

          * Add -R option. Limit the number of requests processed by a connection
            at once. Prevents starving other threads if bulk loading. (Facebook)

2008-07-29 [Version 1.2.6 released]

2008-07-24 [Version 1.2.6-rc1 released]

       * Add support for newer automake (Facebook)

       * DTrace support for Solaris/etc (Trond Norbye)

       * LRU tests (Steve Yen)

       * Handle negative length items properly (Dormando)

       * Don't leave stale data after failed set attempts (Dormando)

       * Fix refcount leaks, which would result in OOM's on all sets
	       (Dormando)

       * Fix buffer overruns (Dustin Sallings, Tomash Brechko)

       * Fix memory corruption with CAS (Dustin Sallings)

2008-06-11

       * Fix -k to work with -d. (reported by Gary Zhu)

2008-03-02 [Version 1.2.5-rc1 released]

       * Add per-item-class tracking of evictions and OOM errors (dormando)

       * Optimize item_alloc() a little (dormando)

       * Give 'SERVER_ERROR out of memory' errors more context (dormando)

       * Enable usage of large memory pages under solaris
         (Trond.Norbye@Sun.COM)

       * Enable UDP by default, clean up server socket code
         (brian@tangent.org)

       * 'noreply' support (Tomash Brechko)

       * IPv6 support, and IPv6 multi-interface support (brian@tangent.org)

       * Add compiler options for Sun Studio compilers with --enable-threads
	     (Trond.Norbye@Sun.COM)

       * Add --enable-64bit for mulitarget platforms (Trond.Norbye@Sun.COM)

       * Use gettimeofday(2) instead of time(2).

       * Make -k option work (Tomash Brechko)

	   * Fix chunk slab alignment (Trond.Norbye@Sun.COM)

2007-12-06 [Version 1.2.4 released]

2007-12-05

       * Fix compilation on panther (JS and Dormando)

	   * More CAS tests (Chris Goffinet)

	   * Final fixes for all 1.2.4 features are in, -rc2 sent out.

2007-11-19 [Version 1.2.4-rc1 released]

2007-11-19  Dormando <dormando@rydia.net>

       * Patch series from Tomash Brechko <tomash.brechko@gmail.com>:
         Minor fixes and optimizations.

       * Patches from Chris, Dustin, and Dormando to fix CAS.

       * Prepping for 1.2.4 release.

2007-11-13  Dormando <dormando@rydia.net>

	* Adjusted patch from js <ebgssth@gmail.com>: Compile on OS X Panther
	  and earlier.

2007-11-12  Steven Grimm  <sgrimm@facebook.com>

	* Patch from Tomash Brechko <tomash.brechko@gmail.com>: Always send
	  "SERVER_ERROR out of memory" when memory exhausted.

2007-10-15  Paul Lindner  <lindner@inuus.com>

	* Patch from David Bremner <bremner@unb.ca> that implements
	  a new option "-a" which takes an octal permission mask
	  (like chmod) sets the permissions on the unix domain socket 
	  (specified by "-s").

2007-10-03 Paul Lindner <lindner@inuus.com>
	* Incorporate "cas" operation developed by Dustin
	  Sallings <dustin@spy.net> This change allows you
	  to do atomic changes to an existing key.

	* Fix for stats.evictions not incrementing
	  when exptime == 0 items are kicked off the cache. 
	  from Jean-Francois BUSTARRET <jfbustarret@wat.tv>.

	* Fix for do_item_cachedump() which was returning
	  an incorrect timestamp.
	
	* Switch to unsigned 64-bit increment/decrement counters
	  from Evan Miller and Dustin Sallings.

	* Add append command support written by Filipe Laborde.
	  Thread safe version plus prepend command from Maxim Dounin
	  <mdounin@mdounin.ru>

	* The memcached-tool script can now display stats.  Patch
	  provided by Dan Christian <dchristian@google.com>

	* Fix for Unix Domain sockets on FreeBSD
	  FreeBSD's sendmsg() requires msg_name in msghdr structure 
	  to be NULL if not used, setting msg_namelen to 0 isn't enough.
	  Patch from Maxim Dounin <mdounin@mdounin.ru>

2007-08-21 Paul Lindner <lindner@inuus.com>
	* Incorporate increment patch from Evan Miller 
	  <emiller@imvu.com> to define increment overflow
	  behavior.

2007-08-07 Leon Brocard <acme@astray.com>

daemon.c
/*    $Header: /cvsroot/wikipedia/willow/src/bin/willow/daemon.c,v 1.1 2005/05/02 19:15:21 kateturner Exp $    */
/*    $NetBSD: daemon.c,v 1.9 2003/08/07 16:42:46 agc Exp $    */
/*-
 * Copyright (c) 1990, 1993
 *    The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if defined __SUNPRO_C || defined __DECC || defined __HP_cc
# pragma ident "@(#)$Header: /cvsroot/wikipedia/willow/src/bin/willow/daemon.c,v 1.1 2005/05/02 19:15:21 kateturner Exp $"
# pragma ident "$NetBSD: daemon.c,v 1.9 2003/08/07 16:42:46 agc Exp $"
#endif

#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "memhc.h"

int daemonize(int nochdir, int noclose)
{
    int fd;

    switch (fork()) {
    case -1:
        return (-1);
    case 0:
        break;
    default:
        _exit(EXIT_SUCCESS);
    }

    if (setsid() == -1)
        return (-1);

    if (nochdir == 0) {
        if(chdir("/") != 0) {
            perror("chdir");
            return (-1);
        }
    }

    if (noclose == 0 && (fd = open("/dev/null", O_RDWR, 0)) != -1) {
        if(dup2(fd, STDIN_FILENO) < 0) {
            perror("dup2 stdin");
            goto err_cleanup;
        }
        if(dup2(fd, STDOUT_FILENO) < 0) {
            perror("dup2 stdout");
            goto err_cleanup;
        }
        if(dup2(fd, STDERR_FILENO) < 0) {
            perror("dup2 stderr");
            goto err_cleanup;
        }

        if(close(fd) < 0) {
            perror("close");
            return (-1);
        }
    }
    return (0);

    err_cleanup:
        close(fd);
        return (-1);
}
